### PR TITLE
Icons bugs/various Improvements

### PR DESCRIPTION
**Pull request by uuilmer**
Was so far able to:

Improve the way I handle Fragment Transactions, which improved performance:

ezgif-6-0e9edea46c35

### CODE REVIEW

**Comment by uuilmer** [comment]
Was able to improve the authenticating UI:

ezgif-6-c7781dfe96c0

**Comment by uuilmer** [comment]
Was able to delete the Search Popup if we resume a ComposeFragment:

ezgif-6-988b3610ab70

**Comment by uuilmer** [comment]
Start/Stop Stream Starts/Stops Spotify:

ezgif-4-a137a979c5e7

File: app/src/main/java/com/example/pix/chat/activities/FriendActivity.java [hunk 1]
```diff
@@ -17,17 +17,20 @@
...
+    public static final String FRIEND_FRAGMENT_TAG = "friend";
+
     @RequiresApi(api = Build.VERSION_CODES.M)
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_friend);
 
         // Check if we took a picture from HomeFragment's ComposeFragment
+        // Set a tag for this Fragment so we can retrieve it if we go to this friend's ProfileFragment
```
Comment (uuilmer) [inline app/src/main/java/com/example/pix/chat/activities/FriendActivity.java:29]: I need to retrieve this FriendFragment in the friend's ProfileFragment so that I may hide the ProfileFragment and show the FriendFragment.

File: app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java [hunk 2]
```diff
@@ -100,11 +101,17 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
+        composeFragment.setEnterTransition(new Explode());
+        setExitTransition(new Explode());
+        composeFragment.setSharedElementReturnTransition(new Explode());
+
         // If we click the camera, go to the ComposeFragment
         ivCamera.setOnClickListener(view1 -> {
             getActivity().getSupportFragmentManager()
                     .beginTransaction()
-                    .replace(R.id.friend_container, new ComposeFragment())
+                    .replace(R.id.friend_container, composeFragment)
```
Comment (uuilmer) [inline app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java:114]: I want to use the same ComposeFragment for this friend to improve transition speed.

File: app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java [hunk 3]
```diff
@@ -143,11 +150,24 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
             Toast.makeText(getContext(), "Error retrieving more chats", Toast.LENGTH_SHORT).show();
         }
 
-        ivProfile.setOnClickListener(view12 -> getParentFragmentManager()
-                .beginTransaction()
-                .addToBackStack("stack")
-                .replace(R.id.friend_container, new ProfileFragment(friend))
-                .commit());
+        Fragment friendProfile = new ProfileFragment(friend);
+        ivProfile.setOnClickListener(view12 -> {
```
Comment (uuilmer) [inline app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java:154]: I am planning to fix these small "view12" and/or missing constants after I merge.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 4]
```diff
@@ -33,18 +49,20 @@
     // we open.
     public static Timer listenerTimer;
     private ImageView ivPlay;
+    public static Subscription<PlayerState> subscription;
```
Comment (uuilmer) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: I want global access to our subscription with a Streamer's Spotify so that I can cancel it below.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 5]
```diff
@@ -268,32 +327,35 @@ public void run() {
...
 
         // If we were listening to a stream via a listenerTimer, stop it
         if (listenerTimer != null) {
             listenerTimer.cancel();
+            listenerTimer = null;
         }
         // If we were streaming as the owner, stop
         if (isOwner) {
-            remote.getPlayerApi().subscribeToPlayerState().cancel();
+            subscription.cancel();
```
Comment (uuilmer) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Here I cancel the global subscription.

File: app/src/main/java/com/example/pix/login/LoginActivity.java [hunk 6]
```diff
@@ -25,21 +31,37 @@
...
     private void checkIfDone() {
         if (loggedIn && authenticated) {
-            Intent i = new Intent(LoginActivity.this, HomeActivity.class);
-            startActivity(i);
-            finish();
+            // If we are missing a permission, request it
+            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED
+                    || ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED
+                    || ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED
+                    || ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED
```
Comment (uuilmer) [inline app/src/main/java/com/example/pix/login/LoginActivity.java:47]: I forgot these last two permissions and later realized they are needed and are why I couldn't change the User's profile picture.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 7]
```diff
@@ -90,23 +108,76 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
+            isPlayingLocally = true;
+            ivPlay.setImageResource(R.drawable.musicroom_pause);
+            rippleBackground.startRippleAnimation();
+        } else {
+            ivPlay.setImageResource(R.drawable.musicroom_play);
         }
+        ivPlay.setOnClickListener(view1 -> {
+            // If we have no Song, make one
+            if (nowPlayingInParse == null) seekSong();
+            else startSync();
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: `{}` please?

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 8]
```diff
@@ -90,23 +108,76 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
         }
 
+        AndroidLikeButton albLike = view.findViewById(R.id.like);
+        likesThisStreamer = Like.checkIfLikes(ParseUser.getCurrentUser(), musicRoom.getUser());
+
+        if (likesThisStreamer == null) {
+            albLike.setCurrentlyLiked(false);
+        } else {
+            albLike.setCurrentlyLiked(true);
+        }
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: If you are the owner here, do you even see this button? If not, no need to have this logic for the owner.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 9]
```diff
@@ -90,23 +108,76 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
-            seekSong();
-        }
-        // Case where there is already a Song playing (This is mainly for listeners)
-        else {
-            setup();
+        // Setup the views based on Parse
+        setup();
+    }
+
+    private void setup() {
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: It seems in this class `MusicRoomFragment` lots of the methods is split into "owner" and not "owner" logic, making the methods long and hard to follow. You should consider create different fragments for these two cases. You can still share some code by having them both extend a base class. So something like "MusicRoomBaseFragment" and "MusicRoomOwnerFragment" and "MusicRoomListenerFragment".

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 10]
```diff
@@ -127,77 +198,50 @@ private void seekSong() {
...
-                            this.cancel();
-                        }
-                    } catch (ParseException e) {
-                        e.printStackTrace();
-                        Toast.makeText(getContext(), "Error looking for new Song!", Toast.LENGTH_SHORT).show();
-                        this.cancel();
-                    }
+        } else {
+            try {
+                musicRoom = q.getFirst();
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Why do you need to do this now? Can't you just set the `musicRoom` when the query was first executed?

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 11]
```diff
@@ -235,15 +283,30 @@ private void startStream(){
                     remote.getPlayerApi().pause();
                 // Every fixed interval, check for updates to Parse
                 listenerTimer = new Timer();
+                // I need to save this Activity, because I won't be able to call getActivity() if we
+                // change from FriendActivity back to HomeActivity, and our listenerTimer would throw an error
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: What does this comment mean?

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java (hunk 4 above)
Comment (xuyueee) [reply app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Not sure why you need to make it public? Do you not only use it in this class? Same question with Timer.

File: app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java [hunk 12]
```diff
@@ -136,11 +157,18 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
         TextView pix = view.findViewById(R.id.profile_pix);
-        pix.setText("" + user.getInt(USER_PIX));
+        // Check every 2 seconds for how many "Pix"(Likes) this person has
+        timer = new Timer();
+        timer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                getActivity().runOnUiThread(() -> pix.setText(Like.getPix(user) + "P"));
+            }
+        }, 0, 2000);
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java]: I'm curious, is this really important such that you need to refresh all the time to get the latest likes?

File: app/src/main/java/com/example/pix/home/models/Like.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    public static final String STREAMER_CODE = "streamer";
+
+    public ParseUser getListener() {
+        return getParseUser("listener");
+    }
+
+    public static Like checkIfLikes(ParseUser listener, ParseUser streamer){
+        ParseQuery<Like> q = ParseQuery.getQuery(Like.class);
+        q.whereEqualTo("listener", listener);
+        q.whereEqualTo("streamer", streamer);
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/home/models/Like.java]: String constants please.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java (hunk 10 above)
Comment (uuilmer) [reply app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: I need to keep running the query to "refresh" the musicroom because at this point it is null, meaning our original query returned no objects because this friend we are attempting to listen to is not streaming. I made it so that you can keep pressing the play button to refresh and try again.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java (hunk 11 above)
Comment (uuilmer) [reply app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: There was actually no need to save the Activity in the variable. I've fixed it.

File: app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java (hunk 12 above)
Comment (uuilmer) [reply app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java]: Would it be better to do less often, or that I don't do it at all, just once when starting the Fragment?

Comment (xuyueee) [reply app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java]: If it's not that important for it to be updated all the time, I'd suggest not having this timer.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java (hunk 10 above)
Comment (xuyueee) [reply app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: I see.

**Comment by xuyueee** [comment]
Feel free to merge after addressing all the comments.
